src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs

```diff
@@ -9,41 +9,44 @@ namespace System.IO.Enumeration
     /// </summary>
     public unsafe ref struct FileSystemEntry
     {
-        // TODO: Unix implementation https://github.com/dotnet/corefx/issues/26715
-        // Inital implementation is naive and not optimized.
-
         internal static void Initialize(
             ref FileSystemEntry entry,
             Interop.Sys.DirectoryEntry directoryEntry,
             bool isDirectory,
             ReadOnlySpan<char> directory,
             string rootDirectory,
-            string originalRootDirectory)
+            string originalRootDirectory,
+            Span<char> pathBuffer)
         {
             entry._directoryEntry = directoryEntry;
             entry._isDirectory = isDirectory;
             entry.Directory = directory;
             entry.RootDirectory = rootDirectory;
             entry.OriginalRootDirectory = originalRootDirectory;
+            entry._pathBuffer = pathBuffer;
+            FileStatus.Initialize(ref entry._status, isDirectory);
         }
 
         internal Interop.Sys.DirectoryEntry _directoryEntry;
-        private FileSystemInfo _info;
+        private FileStatus _status;
         private bool _isDirectory;
+        private Span<char> _pathBuffer;
+        private ReadOnlySpan<char> _fullPath;
 
-        private FileSystemInfo Info
+        private ReadOnlySpan<char> FullPath
         {
             get
             {
-                if (_info == null)
+                if (_fullPath.Length == 0)
                 {
-                    string fullPath = PathHelpers.CombineNoChecks(Directory, _directoryEntry.InodeName);
-                    _info = _isDirectory
-                        ? (FileSystemInfo) new DirectoryInfo(fullPath, fullPath, _directoryEntry.InodeName, isNormalized: true)
-                        : new FileInfo(fullPath, fullPath, _directoryEntry.InodeName, isNormalized: true);
-                    _info.Refresh();
+                    ReadOnlySpan<char> directory = Directory;
+                    directory.CopyTo(_pathBuffer);
+                    _pathBuffer[directory.Length] = Path.DirectorySeparatorChar;
+                    ReadOnlySpan<char> fileName = _directoryEntry.InodeName;
+                    fileName.CopyTo(_pathBuffer.Slice(directory.Length + 1));
+                    _fullPath = _pathBuffer.Slice(0, directory.Length + fileName.Length + 1);
```

Comment:
nit, since the 1 is for a slash not a null, maybe 
`_fullPath = _pathBuffer.Slice(0, directory.Length + 1 + fileName.Length);`